sources/iTermPowerManager.m

```diff
@@ -11,29 +11,52 @@
 
 NSString *const iTermPowerManagerStateDidChange = @"iTermPowerManagerStateDidChange";
 
-@implementation iTermPowerManager
+@implementation iTermPowerManager {
+    CFRunLoopRef _runLoop;
+    CFRunLoopSourceRef _runLoopSource;
+}
+
+static void ITPMpowerSourceChanged(void *context) {
+    iTermPowerManager* pm = (__bridge iTermPowerManager *)(context);
+    
+    CFStringRef source = IOPSGetProvidingPowerSourceType(NULL);
+    if (source){
+        pm.connectedToPower = [@"AC Power" isEqualToString:(__bridge NSString *)(source)];
+    }
+}
 
 + (instancetype)sharedInstance {
-    return [[self alloc] initPrivate];
+    static iTermPowerManager *sharedInstance = nil;
+    static dispatch_once_t onceToken;
+    
+    dispatch_once(&onceToken, ^{
+        sharedInstance = [[self alloc] initPrivate];
+    });
+    
+    return sharedInstance;
 }
 
 - (instancetype)initPrivate {
     self = [super init];
     if (self) {
-        [self updateCharging];
-        [NSTimer it_scheduledTimerWithTimeInterval:5 repeats:YES block:^(NSTimer * _Nonnull timer) {
-            BOOL before = _connectedToPower;
-            [self updateCharging];
-            if (_connectedToPower != before) {
-                [[NSNotificationCenter defaultCenter] postNotificationName:iTermPowerManagerStateDidChange object:nil];
-            }
-        }];
+        _runLoop = CFRunLoopGetMain();
+        _runLoopSource = IOPSCreateLimitedPowerNotification(ITPMpowerSourceChanged,(__bridge void *)(self));
```

Comment:
I think this might  not call post a notification when you plug in. The description of the function says:

````
Returns a CFRunLoopSourceRef that notifies the caller when power source
changes from an unlimited power source (like attached to wall, car, or airplane power), to a limited
power source (like a battery or UPS).
````

Reply:
Yep, tested it. called every time back and forth.